display/d.redraw/main.c

```diff
@@ -0,0 +1,116 @@
+/****************************************************************************
+ *
+ * MODULE:       d.redraw
+ * AUTHOR(S):    Huidae Cho <grass4u gmail.com>
+ *               Based on scripts/d.redraw/d.redraw.py by Martin Landa
+ * PURPOSE:      Redraws the content of currently selected monitor
+ * COPYRIGHT:    (C) 2024 by the GRASS Development Team
+ *
+ *               This program is free software under the GNU General
+ *               Public License (>=v2). Read the file COPYING that
+ *               comes with GRASS for details.
+ *
+ *****************************************************************************/
+
+#include <stdlib.h>
+#include <string.h>
+#include <grass/gis.h>
+#include <grass/spawn.h>
+#include <grass/display.h>
+#include <grass/glocale.h>
+
+#define LINES_SIZE_INC 1024
+#define LINE_LEN       1024
+
+int main(int argc, char **argv)
+{
+    struct GModule *module;
+    const char *mon;
+    char element[GPATH_MAX], cmd_file[GPATH_MAX];
+    FILE *fp;
+    int lines_size, num_lines, num_comment_lines;
+    char **lines, line[LINE_LEN];
+    char **cmd_argv;
+    int i;
+
+    G_gisinit(argv[0]);
+
+    module = G_define_module();
+    G_add_keyword(_("display"));
+    G_add_keyword(_("graphics"));
+    G_add_keyword(_("monitors"));
+    module->description =
+        _("Redraws the content of currently selected monitor.");
+
+    if (G_parser(argc, argv))
+        exit(EXIT_FAILURE);
+
+    if (!(mon = G_getenv_nofatal("MONITOR")))
+        G_fatal_error(_("No graphics device selected. Use d.mon to select "
+                        "graphics device."));
+
+    if (D_open_driver() < 0) {
+        /* first time it runs, add d.redraw to cmd file and exit; this module
+         * will be rerun by the monitor with the display extent (D_open_driver()
+         * will return 0 then) and use that information to constrain rendering
+         */
+        D_close_driver();
+        exit(EXIT_SUCCESS);
+    }
+
+    D_close_driver();
+
+    G_temp_element(element);
+    strcat(element, "/MONITORS/");
+    strcat(element, mon);
+    G_file_name(cmd_file, element, "cmd", G_mapset());
+
+    if (!(fp = fopen(cmd_file, "r")))
+        G_fatal_error(_("Unable to open file '%s' for reading."), cmd_file);
+
+    lines_size = num_lines = num_comment_lines = 0;
+    lines = NULL;
+
+    /* read and save cmd lines; run display commands now */
+    while (G_getl2(line, LINE_LEN, fp)) {
+        /* don't add d.redraw this time */
+        if (strcmp(line, "d.redraw") == 0) {
+            /* remove its comment lines above */
+            num_lines -= num_comment_lines;
+            continue;
+        }
+        if (lines_size == num_lines) {
+            lines_size += LINES_SIZE_INC;
+            lines = G_realloc(lines, sizeof(char *) * lines_size);
+        }
+        lines[num_lines++] = G_store(line);
+
+        if (*line == '#') {
+	    num_comment_lines++;
```

Comment:
@wenzeslaus it works!

Reply:
I'm confused... so we have a new bot that runs clang-format and its formatting is different from the current script?

Reply:
Oh! Never mind. I think I pushed this PR from a laptop without auto-formatting. Just ran it manually.

Reply:
No, clang-format was already failing before I integrated #3483. But now, instead of running and discarding the results, they are changed in place, then the diff is passed to the tool to comment a code suggestion that can be committed directly in the PR, when it is possible to do so.

I updated the branch with main yesterday to show that it was working.

Reply:
I didn't know about the script you added to the comment, I was talking about clang-format version 15 that was in the CI

Reply:
The new workflow takes 30 seconds instead of 20+ mins by the way, so it's a real win